include/picongpu/particles/externalBeam/density/ProbingBeamImpl.hpp

```diff
@@ -0,0 +1,198 @@
+        {
+            static constexpr float_64 get()
+            {
+                return CELL_WIDTH_SI * CELL_HEIGHT_SI;
+            }
+        };
+
+    } // namespace detail
+
+    template<typename T_ParamClass>
+    struct ProbingBeamImpl : public T_ParamClass
+    {
+        using ParamClass = T_ParamClass;
+        using ProbingBeam = typename T_ParamClass::ProbingBeam;
+        // Defines from which side the beam enters the simulation box.
+        using Side = typename T_ParamClass::ProbingBeam::Side;
+
+        static constexpr float_64 photonFluxAtMaxBeamIntensity = ParamClass::photonFluxAtMaxBeamIntensity_SI;
+
+    private:
+        // beam <-> pic index conversions
+
+        static constexpr float_X cellSizeCT[3] = {CELL_WIDTH, CELL_HEIGHT, CELL_DEPTH};
+        static constexpr float_X cellDepth{cellSizeCT[Side::template BeamToSimIdx_t<2u>::value]};
+        //  true if the beam is propagating in a negative direction
+        static constexpr bool reverse = Side::reverse[2];
+
+        // the center of the reduce cell (only Delta t * c in prop direction), relative in cell position
+        static constexpr float_X reducedCellCenterBeam[3]
+            = {0.5_X, 0.5_X, 0.5_X * DELTA_T* SPEED_OF_LIGHT / cellDepth};
+        static constexpr float_X reducedCellCenter_x{reducedCellCenterBeam[Side::template SimToBeamIdx_t<0u>::value]};
+        static constexpr float_X reducedCellCenter_y{reducedCellCenterBeam[Side::template SimToBeamIdx_t<1u>::value]};
+        static constexpr float_X reducedCellCenter_z{reducedCellCenterBeam[Side::template SimToBeamIdx_t<2u>::value]};
+
+
+        // the area of the cell side perpendicular to the propagation
+        static constexpr float_64 CELL_AREA_SI = detail::GetCellAreaSI<Side>::get();
+        static constexpr float_X PHOTONS_IN_A_CELL
+            = static_cast<float_X>(photonFluxAtMaxBeamIntensity * SI::DELTA_T_SI * CELL_AREA_SI);
+        // the photon density in the cell in terms of the base density
+        static constexpr float_X REFERENCE_PHOTON_DENSITY = PHOTONS_IN_A_CELL / CELL_VOLUME / BASE_DENSITY;
+
+    public:
+        template<typename T_SpeciesType>
+        struct apply
+        {
+            using type = ProbingBeamImpl<ParamClass>;
+        };
+        /* Host side constructor
+         *
+         * Provides domain info used for checking if the functor is at a simulation boundary
+         */
+        HINLINE ProbingBeamImpl(uint32_t const& currentStep) : probingBeam_m(), currentStep_m(currentStep)
+        {
+            // TODO: modernize it offset  should be in subGrid already?
+            uint32_t const numSlides = MovingWindow::getInstance().getSlideCounter(currentStep);
+            SubGrid<simDim> const& subGrid = Environment<simDim>::get().SubGrid();
+            DataSpace<simDim> const localCells = subGrid.getLocalDomain().size;
+            globalDomain_m = subGrid.getGlobalDomain();
+            globalDomain_m.offset.y() += numSlides * localCells.y();
```

Comment:
**BUG??** This look wrong, `subGrid.getGlobalDomain().offset` contains already all slides. You account twice for the slides in y direction.